service/policies.go

```diff
@@ -0,0 +1,86 @@
+package service
+
+import (
+	"sync"
+
+	v1 "code.vegaprotocol.io/protos/vega/wallet/v1"
+	"code.vegaprotocol.io/vegawallet/crypto"
+	"github.com/golang/protobuf/jsonpb"
+)
+
+type ConsentConfirmation struct {
+	TxStr    string
+	Decision bool
+}
+
+type ConsentRequest struct {
+	tx            *v1.SubmitTransactionRequest
+	Confirmations chan ConsentConfirmation
+}
+
+func (r *ConsentRequest) String() (string, error) {
+	m := jsonpb.Marshaler{Indent: "    "}
+	marshalledRequest, err := m.MarshalToString(r.tx)
+	return marshalledRequest, err
+}
+
+type Policy interface {
+	Ask(tx *v1.SubmitTransactionRequest) bool
+	GetSignRequestsConfirmations(hash string) chan ConsentConfirmation
+}
+
+type AutomaticConsentPolicy struct {
+	pendingEvents chan ConsentRequest
+}
+
+func NewAutomaticConsentPolicy(pending chan ConsentRequest) Policy {
+	return &AutomaticConsentPolicy{
+		pendingEvents: pending,
+	}
+}
+
+func (p *AutomaticConsentPolicy) Ask(tx *v1.SubmitTransactionRequest) bool {
+	return true
+}
+
+func (p *AutomaticConsentPolicy) GetSignRequestsConfirmations(hash string) chan ConsentConfirmation {
+	return nil
+}
+
+type ExplicitConsentPolicy struct {
+	pendingEvents    chan ConsentRequest
+	AllConfirmations sync.Map
+}
+
+func NewExplicitConsentPolicy(pending chan ConsentRequest) Policy {
+	return &ExplicitConsentPolicy{
+		pendingEvents:    pending,
+		AllConfirmations: sync.Map{},
+	}
+}
+
+func (p *ExplicitConsentPolicy) GetSignRequestsConfirmations(hash string) chan ConsentConfirmation {
+	confirmations, ok := p.AllConfirmations.Load(hash)
+	if ok {
+		return confirmations.(chan ConsentConfirmation)
+	}
+	return nil
+}
+
+func (p *ExplicitConsentPolicy) Ask(tx *v1.SubmitTransactionRequest) bool {
+	txHash := crypto.AsSha256(tx)
+	confirmations := make(chan ConsentConfirmation)
+	p.AllConfirmations.Store(txHash, confirmations)
+	p.pendingEvents <- ConsentRequest{tx: tx, Confirmations: confirmations}
+
+	for c := range confirmations {
```

Comment:
We're only ever expecting one thing through this channel now so we don't need the loop. And we should either trust our code and not do the hash check, or keep the hash check and panic if its mismatches because if it does theres a bug in our code.

Reply:
done